src/raftstore/store/msg.rs

```diff
@@ -18,11 +18,81 @@ use std::fmt;
 use kvproto::raft_serverpb::RaftMessage;
 use kvproto::raft_cmdpb::{RaftCmdRequest, RaftCmdResponse};
 use kvproto::metapb::RegionEpoch;
+
 use raft::SnapshotStatus;
 use util::escape;
 
-pub type Callback = Box<FnBox(RaftCmdResponse) + Send>;
-pub type BatchCallback = Box<FnBox(Vec<Option<RaftCmdResponse>>) + Send>;
+use super::RegionSnapshot;
+
+#[derive(Debug)]
+pub struct ReadResponse {
```

Comment:
To avoid ambiguity, I think it is better to use Snap instead of Read here, because now we can only get a snapshot from raftstore or local read thread.

Btw, I even think we should rename local read thread to local snap thread.

Reply:
It's for read-only requests. Get, Snap, Status, etc.

Reply:
For me, read is just read, not others, especially not Status.

Reply:
So what's read exactly? Is it a read to get the cluster status? Is it a read to get a key? Is it a read to make a snapshot (but don't get any keys)?

Reply:
Read means read-only, it does not mutate the state.

Reply:
I suggest using different type here, read, snap and status.

Reply:
I agree with read-only too. There will be duplicated types that takes exact the same action but with different name if use different type for each request.

Reply:
read will read a value, snap will get a snapshot, and status only get status, can they be combined in a same request?